src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -391,10 +412,20 @@ struct
 
   let join a b =
     let join_d ad bd =
-      (*use copy of ad because result is later saved in there*)
-      let ad = Array.copy ad in
-      (*This is the table which is later grouped*)
-      let table = BatList.map2i (fun i (ai, aj) (bi,bj) -> (i, Z.(aj - bj), (ai, aj), (bi,bj))) (Array.to_list ad) (Array.to_list bd) in
+      (* joinfunction handles the dirty details of performing an "inner join" on the lhs of both bindings;
+         in the resulting binding, the lhs is then mapped to values that are later relevant for sorting/grouping, i.e.
+         - the difference of both offsets
+         - rhs1
+         - rhs2 
+           however, we have to account for the sparseity of EConj maps by manually patching holes with default values *)
+      let joinfunction lhs rhs1 rhs2 = match rhs1, rhs2 with 
+        | Some (ai,aj),Some (bi,bj) -> Some (Z.(aj - bj),(ai,aj),(bi,bj))           (* this is explicitely what we want *)  
+        | None, Some (bi,bj)        -> Some (Z.neg bj   ,Rhs.var_zero lhs,(bi,bj))  (* account for the sparseity of binding 1 *)
+        | Some (ai,aj), None        -> Some (aj         ,(ai,aj),Rhs.var_zero lhs)  (* account for the sparseity of binding 2 *)
+        | _,_                       -> None  (* no binding for lhs in both maps is replicated implicitely in a sparse result map *)
+      in
+      let flatten (a,(b1,b2,b3)) = (a,b1,b2,b3) in
+      let table = List.map flatten @@ EConj.IntMap.bindings @@ EConj.IntMap.merge joinfunction (snd ad) (snd bd) in
```

Comment:
Seems like `joinfunction` could already do the `flatten` logic in the same go.

Reply:
I do not really see how I can do that, due to the flatten working on the tuple that gets as snd element the triple produced by the joinfunction, but maybe I missed something here?

Reply:
I mean, the first case of `joinfunction` could return
```ocaml
Some (lhs, Z.(aj - bj),(ai,aj),(bi,bj))
```
Just put the key (`lhs`) already into the result quadruple, just like the `flatten` on `bindings` would.

Although maybe this won't make much of a difference if `table` still needs to be a list. Then bindings (or at least values) still need to be transformed into a list.

Reply:
Yeah, table needs to be a list for `.group` to happen later on in the code... still, I like your idea, getting rid of flatten.